File: TestSuite.cpp
#include "TestSuite.hpp"
#include "AVLTree.hpp"
#include <iostream>
#include <cmath>

using namespace std;

// this function runs every test in order and prints the results.
void TestSuite::runAllTests() {
    cout << "\n===== Running Test Suite =====" << endl;
    testInsert();
    testDelete();
    testSearch();
    testBalancing();
    testCollisionHandling();
    cout << "===== Test Suite Complete =====" << endl;
}

// this function tests that inserting an ip makes it findable through the hash table.
void TestSuite::testInsert() {
    LogAnalyzer analyzer;
    analyzer.processConnection("1.1.1.1", "2026-07-29 09:00", 80, false);

    IPRecord* result = analyzer.lookupIP("1.1.1.1");
    if (result != nullptr && result->getIP() == "1.1.1.1") {
        cout << "testInsert: PASS" << endl;
    } else {
        cout << "testInsert: FAIL" << endl;
    }
}

// this function tests that removing a risk score from the tree actually removes it.
void TestSuite::testDelete() {
    AVLTree tree;
    IPRecord record("2.2.2.2");
    record.increaseRisk(5);
    tree.insert(&record);

    bool foundBefore = tree.contains(5);
    tree.remove(5);
    bool foundAfter = tree.contains(5);

    if (foundBefore == true && foundAfter == false) {
        cout << "testDelete: PASS" << endl;
    } else {
        cout << "testDelete: FAIL" << endl;
    }
}

// this function tests that searching for an ip that was never added returns null.
void TestSuite::testSearch() {
    LogAnalyzer analyzer;
    IPRecord* result = analyzer.lookupIP("9.9.9.9");

    if (result == nullptr) {
        cout << "testSearch: PASS" << endl;
    } else {
        cout << "testSearch: FAIL" << endl;
    }
}

// this function tests that the tree stays balanced after several inserts.
// a balanced tree with n nodes should have a height close to log base two of n.
void TestSuite::testBalancing() {
    AVLTree tree;
    IPRecord records[7] = {
        IPRecord("10.0.0.1"), IPRecord("10.0.0.2"), IPRecord("10.0.0.3"),
        IPRecord("10.0.0.4"), IPRecord("10.0.0.5"), IPRecord("10.0.0.6"),
        IPRecord("10.0.0.7")
    };

    // give each record an increasing risk score before inserting.
    for (int i = 0; i < 7; i++) {
        records[i].increaseRisk(i + 1);
        tree.insert(&records[i]);
    }

    int height = tree.getTreeHeight();
    int expectedMaxHeight = static_cast<int>(log2(7)) + 2;

    if (height <= expectedMaxHeight) {
        cout << "testBalancing: PASS (height " << height << ")" << endl;
    } else {
        cout << "testBalancing: FAIL (height " << height << " too tall)" << endl;
    }
}

// this function tests that adding the same ip twice updates the existing record
// instead of creating a duplicate entry in the hash table.
void TestSuite::testCollisionHandling() {
    LogAnalyzer analyzer;
    analyzer.processConnection("3.3.3.3", "2026-07-29 09:00", 80, false);
    analyzer.processConnection("3.3.3.3", "2026-07-29 09:05", 22, true);

    int count = analyzer.getTrackedCount();
    IPRecord* result = analyzer.lookupIP("3.3.3.3");

    if (count == 1 && result != nullptr && result->getAttemptCount() == 2) {
        cout << "testCollisionHandling: PASS" << endl;
    } else {
        cout << "testCollisionHandling: FAIL" << endl;
    }
}

File: web/src/utils/testSuite.ts
import { LogAnalyzer } from "../models/LogAnalyzer";
import { AVLTree } from "../models/AVLTree";
import { IPRecord } from "../models/IPRecord";

// this type represents the result of a single test.
// the ui will use this to show pass or fail with a label.
export interface TestResult {
  name: string;
  passed: boolean;
  detail: string;
}

// this function tests that inserting an ip makes it findable through the hash table.
function testInsert(): TestResult {
  const analyzer = new LogAnalyzer();
  analyzer.processConnection("1.1.1.1", "2026-07-29 09:00", 80, false);

  const result = analyzer.lookupIP("1.1.1.1");
  const passed = result !== null && result.getIP() === "1.1.1.1";

  return { name: "testInsert", passed, detail: "" };
}

// this function tests that removing a risk score from the tree actually removes it.
function testDelete(): TestResult {
  const tree = new AVLTree();
  const record = new IPRecord("2.2.2.2");
  record.increaseRisk(5);
  tree.insert(record);

  const foundBefore = tree.contains(5);
  tree.remove(5);
  const foundAfter = tree.contains(5);

  const passed = foundBefore === true && foundAfter === false;

  return { name: "testDelete", passed, detail: "" };
}

// this function tests that searching for an ip that was never added returns null.
function testSearch(): TestResult {
  const analyzer = new LogAnalyzer();
  const result = analyzer.lookupIP("9.9.9.9");

  const passed = result === null;

  return { name: "testSearch", passed, detail: "" };
}

// this function tests that the tree stays balanced after several inserts.
// a balanced tree with n nodes should have a height close to log base two of n.
function testBalancing(): TestResult {
  const tree = new AVLTree();
  const records: IPRecord[] = [];

  for (let i = 0; i < 7; i++) {
    const record = new IPRecord(`10.0.0.${i + 1}`);
    record.increaseRisk(i + 1);
    records.push(record);
    tree.insert(record);
  }

  const height = tree.getTreeHeight();
  const expectedMaxHeight = Math.floor(Math.log2(7)) + 2;
  const passed = height <= expectedMaxHeight;

  return { name: "testBalancing", passed, detail: `height ${height}` };
}

// this function tests that adding the same ip twice updates the existing record
// instead of creating a duplicate entry in the hash table.
function testCollisionHandling(): TestResult {
  const analyzer = new LogAnalyzer();
  analyzer.processConnection("3.3.3.3", "2026-07-29 09:00", 80, false);
  analyzer.processConnection("3.3.3.3", "2026-07-29 09:05", 22, true);

  const count = analyzer.getTrackedCount();
  const result = analyzer.lookupIP("3.3.3.3");

  const passed = count === 1 && result !== null && result.getAttemptCount() === 2;

  return { name: "testCollisionHandling", passed, detail: "" };
}

// this function runs every test in order and returns all the results.
// the ui calls this and renders the results list instead of printing to console.
export function runAllTests(): TestResult[] {
  return [
    testInsert(),
    testDelete(),
    testSearch(),
    testBalancing(),
    testCollisionHandling(),
  ];
}